rfd/0199-hardware-key-agent.md

```diff
@@ -0,0 +1,574 @@
+cached, the Teleport client will provide the PIN to the hardware key without
+prompting the user again.
+
+#### Cluster Auth Preference
+
+To enable PIN caching for Teleport clients, set `cap.hardware_key.pin_cache_timeout`
+to the desired timeout duration:
+
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  ...
+  hardware_key:
+    # pin_cache_timeout is the amount of time that Teleport clients will cache
+    # the user's hardware key (PIV) PIN. The timeout countdown is started when 
+    # the PIN is stored and is not extended by subsequent accesses. This timeout
+    # can not exceed 1 hour. When empty or 0, the pin will not be cached.
+    pin_cache_timeout: 15m
+```
+
+Teleport clients will retrieve this setting through `/webapi/ping`, which is
+cached by the client alongside other cluster settings.
+
+#### `PinCachingPrompt` pseudo-code
+
+PIN caching will be implemented at the level of the PIN prompt itself, holding
+the PIN in memory and only prompting for PIN once the cache timeout duration
+has elapsed.
+
+```go
+// Pseudo-code
+
+// PinCachingPrompt is a HardwareKeyPrompt wrapped with PIN caching.
+type PinCachingPrompt struct {
+  keys.HardwareKeyPrompt
+
+  // cluster-configurable timeout duration
+  PinCacheTimeout time.Duration
+
+  cachedPIN       string
+  cachedPINExpiry time.Time
+}
+
+func (p *PinCachingPrompt) AskPIN(ctx context.Context, requirement keys.PINPromptRequirement) (string, error) {
+  pin := p.getCachedPIN()
+  if pin != "" {
+    return pin, nil
+  }
+
+  pin = p.HardwareKeyPrompt.AskPIN()
+  p.setCachedPIN(pin)
+  return pin, nil
+}
+
+func (p *PinCachingPrompt) getCachedPIN() (string) {
+  if p.cachedPIN == "" {
+    return ""
+  }
+  
+  if time.Now().Before(p.cachedPINExpiry) {
+    return p.cachedPIN
+  }
+
+  if time.Now().After(p.cachedPINExpiry) {
+    p.cachedPIN = ""
+    return ""
+  }
+}
+
+func (p *PinCachingPrompt) setCachedPIN(pin string) {
+  p.cachedPIN = pin
+  p.cachedPINExpiry = time.Now().Add(p.PinCacheTimeout)
+}
+```
+
+Note: like a normal client, the agent client will use the `PinCachingPrompt`
+when configured for the cluster, so it will naturally cache the PIN for all of
+its dependent clients.
+
+### Hardware Key Agent
+
+The hardware key agent will provide two functionalities:
+
+1. Fetching info for a specific hardware private key, including the public
+key, private key policy, and attestation statement.
+1. Signing with a specific hardware private key
+
+Note: this is the base functionality needed for dependent clients to implement
+the `crypto.Signer` interface and Teleport's `keys.HardwareSigner` interface.
+
+The agent will be served as a [gRPC](#hardwarekeyagentservice) service on a unix
+socket, `$TEMP/.Teleport-PIV/agent.sock`, with [basic TLS](#security).
+
+#### `$TEMP/.Teleport-PIV/agent.sock`
+
+`$TEMP` here depends on the client OS. We will use `os.TempDir` to get the
+correct temp directory for the OS.
```

Comment:
`os.TempDir` on macOS returns a user-specific directory under something like `/var/folders/kz/<some random characters>/T/`. I guess that doesn't matter much because of the file permissions you list below, but later in the RFD I found references to `/tmp` which is available to all users.